### ISSUE TITLE
PicklingError when autodocumenting a class with a dynamically created parent

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1302]
In IPython, we're using a copy of the six with_metaclass function for Python 2 & 3 compatibility. This dynamically creates a `NewBase` class, which the real class will inherit from. In our docs, we use the `autoclass` directive with the `:show-inheritance:` option. This fails with the error given below.

The failure appears to be new in Sphinx 1.2 - I don't see it with 1.1.3.

[This issue](http://code.tabo.pe/django-treebeard/issue/60/docs-building-process-breaks-with-django) reported for the Django-treebeard library appears to be the same.

```
# Sphinx version: 1.2b3
# Python version: 2.7.5+
# Docutils version: 0.11 release
# Jinja2 version: 2.7.1
# Loaded extensions:
#   sphinx.ext.graphviz from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/ext/graphviz.pyc
#   github from /home/takluyver/Code/ipython/docs/sphinxext/github.pyc
#   sphinx.ext.autodoc from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/ext/autodoc.pyc
#   sphinx.ext.intersphinx from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/ext/intersphinx.pyc
#   sphinx.ext.doctest from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/ext/doctest.pyc
#   matplotlib.sphinxext.mathmpl from /usr/lib/pymodules/python2.7/matplotlib/sphinxext/mathmpl.pyc
#   matplotlib.sphinxext.plot_directive from /usr/lib/pymodules/python2.7/matplotlib/sphinxext/plot_directive.pyc
#   sphinx.ext.inheritance_diagram from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/ext/inheritance_diagram.pyc
#   matplotlib.sphinxext.only_directives from /usr/lib/pymodules/python2.7/matplotlib/sphinxext/only_directives.pyc
#   numpydoc from /home/takluyver/Code/ipython/docs/sphinxext/numpydoc.pyc
#   IPython.sphinxext.ipython_console_highlighting from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/IPython/sphinxext/ipython_console_highlighting.pyc
#   IPython.sphinxext.ipython_directive from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/IPython/sphinxext/ipython_directive.pyc
#   sphinx.ext.oldcmarkup from /home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/ext/oldcmarkup.pyc
Traceback (most recent call last):
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/cmdline.py", line 246, in main
    app.build(force_all, filenames)
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/application.py", line 212, in build
    self.builder.build_update()
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 214, in build_update
    'out of date' &#37; len(to_build))
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 234, in build
    purple, length):
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 134, in status_iterator
    for item in iterable:
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/environment.py", line 470, in update_generator
    self.read_doc(docname, app=app)
  File "/home/takluyver/.virtualenvs/ipy-trunk/local/lib/python2.7/site-packages/sphinx/environment.py", line 689, in read_doc
    pickle.dump(doctree, f, pickle.HIGHEST_PROTOCOL)
PicklingError: Can't pickle <class 'IPython.utils.traitlets.NewBase'>: attribute lookup IPython.utils.traitlets.NewBase failed
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1302
- Originally reported by: Thomas Kluyver
- Originally created at: 2013-11-04T21:28:40.238

**Comment by shimizukawa** [comment]
_From Thomas Kluyver on 2013-11-13 20:43:39+00:00_

I've traced this to https://bitbucket.org/birkenfeld/sphinx/pull-request/25 - it only hits when an inheritance diagram includes a class with a dynamically defined parent.

In the inheritance diagram, class_info is now stored as a dict, keyed by the class objects themselves, whereas previously only the values from that dict were stored. That means that pickling now saves the class objects, not just the names.

A workaround that suffices for my case is to give the dynamically defined class a name beginning with underscore, e.g. `_NewBase`. Then it will be skipped over as a private base class, as long as you don't set the `private-bases` flag in the inheritance diagram directive.

**Comment by shimizukawa** [comment]
_From Jason Penney on 2014-01-15 20:40:02+00:00_

I'm seeing the same issue. It seems this same issue was fixed some time ago (#510), and 1.2 reintroduces the error.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-19 10:35:32+00:00_

Thanks for the report!

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-19 10:35:33+00:00_

Closes #1302: Fix regression in :mod:`sphinx.ext.inheritance_diagram` when documenting classes that can't be pickled.

→ <<cset 3f7ecbde9678c7b52ef15d133c8b291468fe38f2>>
